### PR TITLE
SI-4492 More informative error when class not found on classpath

### PR DESCRIPTION
**Pull request by adriaanm**
Sorry, I don't have time to add a test for the module advice part of the error message.
Our classpath handling needs to be fixed to do this reliably anyway.

I discovered another bug in our launcher script (#3630) while testing it manually...

### CODE REVIEW

**Comment by retronym** [comment]
Perhaps you could paste a transcript of the manual test in lieu of an automated one?

The error message does look nicer.

**Comment by adriaanm** [comment]
Sure -- I'm looking into a nice to reproduce approach, because I had to manually remove scala-xml from the TOOL_CLASSPATH in the runner script....

File: src/reflect/scala/reflect/internal/pickling/UnPickler.scala [hunk 1]
```diff
@@ -229,6 +229,19 @@ abstract class UnPickler {
...
+        def moduleAdvice(missing: String): String = {
+          val module =
+            if      (missing.startsWith("scala.xml"))                Some(("org.scala-lang.modules", "scala-xml"))
+            else if (missing.startsWith("scala.util.parsing"))       Some(("org.scala-lang.modules", "scala-parser-combinators"))
+            else if (missing.startsWith("scala.swing"))              Some(("org.scala-lang.modules", "scala-swing"))
+            else if (missing.startsWith("scala.util.continuations")) Some(("org.scala-lang.plugins", "scala-continuations-library"))
+            else None
+
+          (module map { case (group, art) =>
+            s"""\n(it looks like the $art module is missing; try adding a dependency on:\n "$group" %% "$art" % "(version)")"""
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/pickling/UnPickler.scala]: Perhaps this is being pedantic, but this message is SBT specific. How about a link to a modularization guide on http://docs.scala-lang.org where we can evolve the content?

**Comment by retronym** [comment]
You could probably go for a `run` test where you can tweak the settings/classpath manually. In fact, the check files you touched are from similar `run-neg` tests.

File: src/reflect/scala/reflect/internal/pickling/UnPickler.scala (hunk 1 above)
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/pickling/UnPickler.scala]: I guess I could `s/%+/:/g`, based on http://search.maven.org/#artifactdetails%7Corg.scala-lang.modules%7Cscala-xml_2.11.0-RC1%7C1.0.0%7Cbundle

Do we have a url for that migration guide. We should, by now, I guess...

Comment (gkossakowski) [inline src/reflect/scala/reflect/internal/pickling/UnPickler.scala]: Good point on migration guide. The executable part of migration guide is here: https://github.com/scala/scala-module-dependency-sample

What would be in the guide in addition to the sample code we already have?

Comment (adriaanm) [inline src/reflect/scala/reflect/internal/pickling/UnPickler.scala]: There's a tiny piece of advice at: https://github.com/scala/scala/wiki/Scala-2.11#xml,
which another error message links to. Once we have the url for the migration guide, I can update the PR and also change the url in Contexts.

File: test/files/run/t6440b.check [hunk 2]
```diff
@@ -1,4 +1,5 @@
-pos: NoPosition bad symbolic reference. A signature in U.class refers to type T
-in package pack1 which is not available.
-It may be completely missing from the current classpath, or the version on
-the classpath might be incompatible with the version used when compiling U.class. ERROR
+pos: NoPosition bad symbolic reference to pack1.T encountered in class file 'U.class'.
```
Comment (retronym) [inline test/files/run/t6440b.check:1]: Still `NoPosition` here. Expected?

Comment (adriaanm) [inline test/files/run/t6440b.check:1]: I guess this one isn't rewrapped by the code in typedSelectInternal

File: src/reflect/scala/reflect/internal/pickling/UnPickler.scala (hunk 1 above)
Comment (adriaanm) [inline src/reflect/scala/reflect/internal/pickling/UnPickler.scala]: I propose to create an overview at http://docs.scala-lang.org/overviews/core/scala-2.11.html

File: test/files/run/t6440b.check (hunk 2 above)
Comment (adriaanm) [inline test/files/run/t6440b.check:1]: the missing requirement is triggered by deleting the class file for the class T, whose absence is felt by U, which extends it

Comment (retronym) [inline test/files/run/t6440b.check:1]: I wonder if we can always use something like:

``` scala
val issuePos = errorPos.orElse( { devWarning("Que?"); context.tree.pos })
```

File: src/reflect/scala/reflect/internal/pickling/UnPickler.scala (hunk 1 above)
Comment (retronym) [inline src/reflect/scala/reflect/internal/pickling/UnPickler.scala]: That URL SGTM. We could leave the messages a bit loosely coupled ("See the modularization section in http://")

Comment (adriaanm) [inline src/reflect/scala/reflect/internal/pickling/UnPickler.scala]: I'll use the scala-xml anchor, which won't break the link if we somehow mess it up, but I'll submit a PR with a placeholder now.

**Comment by adriaanm** [comment]
Ok, I think I'm going to do the minimal updates discussed above, and defer the rest to the post-2.11.0-final era.

**Comment by adriaanm** [comment]
Argh, has anybody else had their commit messages eaten by sublime text? It dropped everything but the first line...

**Comment by gkossakowski** [comment]
I never had that.

File: test/files/run/t6440b.check (hunk 2 above)
Comment (adriaanm) [inline test/files/run/t6440b.check:1]: I added a TODO in the code.

**Comment by adriaanm** [comment]
review by @gkossakowski

File: src/compiler/scala/tools/nsc/typechecker/Contexts.scala [hunk 3]
```diff
@@ -544,6 +544,8 @@ trait Contexts { self: Analyzer =>
       if (checking) onTreeCheckerError(pos, msg) else unit.error(pos, msg)
 
     @inline private def issueCommon(err: AbsTypeError)(pf: PartialFunction[AbsTypeError, Unit]) {
+      // TODO: are errors allowed to have pos == NoPosition??
```
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/typechecker/Contexts.scala:547]: What does it mean allowed here? Will they crash the compiler, will it break tools?

I know that Eclipse does handle errors with no position. They are reported at the top of the source file.

Comment (adriaanm) [inline src/compiler/scala/tools/nsc/typechecker/Contexts.scala:547]: I meant, can we enforce the invariant `pos != NoPosition`, or maybe use @retronym's suggestion of replacing `NoPosition` by `context.tree.pos`

Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/typechecker/Contexts.scala:547]: Ok

**Comment by gkossakowski** [comment]
LGTM

**Comment by adriaanm** [comment]
Thanks for the reviews, guys. I will do my best to polish this area a little more next time I hit the TODO trail.

**Pull request merged**